developer/wip/for-contributors/data-plane/custom-data-plane.md

```diff
@@ -0,0 +1,252 @@
+
+```json
+{
+    "@context": {
+        "@vocab": "https://w3id.org/edc/v0.0.1/ns/"
+    },
+    "counterPartyAddress": "{{PROVIDER_DSP}}/api/dsp",
+    "connectorId": "{{PROVIDER_ID}}",
+    "contractId": "{{CONTRACT_ID}}",
+    "dataDestination": {
+        "type": "HttpData",
+        "baseUrl": "{{RECEIVER_URL}}"
+    },
+    "protocol": "dataspace-protocol-http",
+    "transferType": "HttpData-PUSH"
+}
+```
+
+The simplest `handlePush` function would need to fetch data from the source `baseUrl` and send the result to the sink `baseUrl`.
+
+ A naive implementation may look like this:
+
+```javascript
+async function handlePush(req, res) {
+    res.send({
+        "@context": {
+            "edc": "https://w3id.org/edc/v0.0.1/ns/"
+        },
+        "@type": "DataFlowResponseMessage"
+    });
+
+    const { sourceDataAddress, destinationDataAddress } = req.body;
+
+    const sourceUrl = getBaseUrl(sourceDataAddress);
+    const destinationUrl = getBaseUrl(destinationDataAddress);
+
+    const response = await fetch(sourceUrl);
+
+    await fetch(destinationUrl, {
+        "method": "POST",
+        body : await response.text()
+    });
+}
+```
+
+First we acknowledge the [Control Plane](../contributor-handbook.md#2-the-control-plane) by sending a `DataFlowResponseMessage` as response. 
+
+Then we transfer the data from `sourceUrl` to `destinationUrl`.
+
+> The `getBaseUrl` is an utility function that extract the `baseUrl` from the `DataAddress`.
+ 
+Implementors should keep track of `DataFlowStartMessage`s in some persistent storage system in order to fulfill subsequent `DPS` messages on the same transfer id ([e.g. SUSPEND and TERMINATE](#22-suspend-and-terminate)).
+
+For example in the streaming case, implementors may track the opened streaming channels, which could be terminated on-demand or by the [policy monitor](../control-plane/policy-monitor.md).
+
+
+### 2.1.2 `PULL`
+
+When receiving a `DataFlowStartMessage` in a `PULL` scenario there is no direct transfer to be handled by the data plane. Based on the `sourceDataAddress` in the `DataFlowStartMessage` a custom data plane implementation should create another `DataAddress` containing all the information required for the data transfer:
+
+
+```javascript
+async function handlePull(req, res) {
+    const { sourceDataAddress } = req.body;
+    const { dataAddress } = await generateDataAddress(sourceDataAddress);
+
+    const response = {
+        "@context": {
+            "edc": "https://w3id.org/edc/v0.0.1/ns/"
+        },
+        "@type": "DataFlowResponseMessage",
+        "dataAddress": dataAddress
+    };
+    res.send(response);
+}
+```
+
+We will not implement the `generateDataAddress` function, as it may vary depending on the use case. But at high level a `generateDataAddress` should generate a `DataAddress` in DSP format that contains useful information for the consumer for fetching the data: `endpoint`, `endpointType` and custom extensible properties `endpointProperties`. 
```

Comment:
```suggestion
We will not implement the `generateDataAddress` function, as it may vary depending on the use case. But at the high level a `generateDataAddress` should generate a `DataAddress` in DSP format that contains useful information for the consumer for fetching the data: `endpoint`, `endpointType` and custom extensible properties `endpointProperties`. 
```